### PR TITLE
[3006.x] Fix env

### LINKED ISSUE #65002: [BUG] After upgrading to 3006.2 gitfs states 'remote ref does not exist' for ext_pillar
**Description**
After upgrading our salt-masters from 3006.1 to 3006.2 the master is unable to checkout our ext_pillar git with the following error:

~~~
[ERROR   ] Failed to checkout master from git_pillar remote '__env__ https://gitlab.com/my/salt-pillar.git': remote ref does not exist
~~~

Downgrading the salt-master, salt-minion and salt packages back to 3006.1 immediately solves the issue.

**Setup**

Please be as specific as possible and give set-up details.

- [x] on-prem machine
- [x] VM (Virtualbox, KVM, etc. please specify)
- [x] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [x] classic packaging
- [ ] onedir packaging
- [ ] used bootstrap to install

The masters run on a CentOS 9 Stream machine with the latest patches (see below).

**Steps to Reproduce the behavior**
 
This is the snipped from the master conf regarding the ext_pillar that causes the issue:
~~~yaml
ext_pillar:
  - git:
    - __env__ https://gitlab.com/my/salt-pillar.git:
      - root: .
      - user: gitlab-deploy-token
      - password: **********
      - fallback: master
~~~

Running the master in debug mode prints the following:
~~~
salt-master[691168]: [DEBUG   ] Current fetch URL for git_pillar remote '__env__ https://gitlab.com/my/salt-pillar.git': https://gitlab.com/my/salt-pillar.git (desired: https://gitlab.com/my/salt-pillar.git)
salt-master[691168]: [DEBUG   ] Current refspecs for git_pillar remote '__env__ https://gitlab.com/my/salt-pillar.git': ['+refs/heads/*:refs/remotes/origin/*', '+refs/tags/*:refs/tags/*'] (desired: ['+refs/heads/*:refs/remotes/origin/*', '+refs/tags/*:refs/tags/*'])
salt-master[691168]: [DEBUG   ] Current http.sslVerify for git_pillar remote '__env__ https://gitlab.com/my/salt-pillar.git': true (desired: true)
salt-master[691168]: [ERROR   ] Failed to checkout master from git_pillar remote '__env__ https://gitlab.com/my/salt-pillar.git': remote ref does not exist
~~~

The error message originates from [this line of code](https://github.com/saltstack/salt/blob/8f750fa7ae115c48c16ca03ef3f16e4ba76f921c/salt/utils/gitfs.py#L1799C18-L1799C18). Debugging showed the problem to be that there are no references retrieved at [this point](https://github.com/saltstack/salt/blob/8f750fa7ae115c48c16ca03ef3f16e4ba76f921c/salt/utils/gitfs.py#L1632C11-L1632C11) and such the `refs` list being empty.

We also reference other repositories in the ext_pillar block like this and they work fine.
~~~yaml
  - git:
    - main https://gitlab.com/my-other/salt-pillar.git:
      - root: .
      - user: other-pillars
      - password: *********
      - env: onpremises
~~~

**Expected behavior**
We expect the master to checkout the pillar repository without errors like in version 3006.1

**Versions Report**
<details><summary>salt --versions-report</summary>
Masters and Minions have the same version and also run on similarly configured VMs (same OS).

```yaml
Salt Version:
          Salt: 3006.2

Python Version:
        Python: 3.10.12 (main, Aug  3 2023, 21:47:10) [GCC 11.2.0]

Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.2
       libgit2: 1.6.4
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: 1.12.2
  python-gnupg: 0.4.8
        PyYAML: 6.0.1
         PyZMQ: 23.2.0
        relenv: 0.13.3
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: centos 9
        locale: utf-8
       machine: x86_64
       release: 5.14.0-331.el9.x86_64
        system: Linux
       version: CentOS Stream 9
```
</details>

### PR DESCRIPTION
**Pull request by cmcmarrow**
### What does this PR do?
fix `__env__`

Changes: 
* Place remotes branch/tag in its own folder `../salt/../hash/branch`
  * This includes locks and links
* Clear gitfs cache on salt version change 
  * This is so people don't need to manually clear salt cache when upgrading
* Add salt working dir `../salt/../work/hash/branch`
  * This is so we decrease the chances of stepping on git
* add `fetch_on_fail` `checkout(self, fetch_on_fail=True)`
  * This is so gitpillar does not need worry if a fetch is required when working with dynamic branches
* Fix `clean_old_remotes` 
* Fix fetch for `__env__`
  * add fetch_request - when `Gitbase.fetch_remotes` is called the current dynamic branch will fetch and all its sister branches will have a fetch_request file placed in its salt working dir. When sister branch inits or goes to checkout it will see the fetch_request and run `fetch`
* Tests
  
^ fixs
`salt-call pillar.data pillarenv=foo`
`salt-run git_pillar.update`
  
 TODO later PR:
* remove old remotes branch's/tags
  * Make a timestamp file every time a gitprovier object is made in remote branch
  * `clear_old_remotes()` will remove any remote branch that has out lived its time to live.
* Move repo locks to salt working dir

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/65002
Fixes: https://github.com/saltstack/salt/issues/64999

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

File: salt/utils/gitfs.py [hunk 1]
```diff
@@ -2504,48 +2536,42 @@ def init_remotes(
...
+                "Unable to remove old %s remote cachedir %s: %s",
+                self.role,
+                cache_dir,
+                exc,
+            )
+            return False
+        log.debug("%s removed old cachedir %s", self.role, cache_dir)
+        return True
+
+    def _iter_remote_hashs(self):
```
Comment (s0undt3ch) [inline salt/utils/gitfs.py:2553]:
```suggestion
    def _iter_remote_hashes(self):
```
?

File: salt/utils/gitfs.py [hunk 2]
```diff
@@ -2504,48 +2536,42 @@ def init_remotes(
...
-        for fdir in failed:
-            to_remove.remove(fdir)
-        ret = bool(to_remove)
-        if ret:
+        change = False
+        # Remove all hash dirs not part of this group
+        remote_set = {r.get_cache_basehash() for r in self.remotes}
+        for item in self._iter_remote_hashs():
+            if item not in remote_set:
+                change = change or self._remove_cache_dir(
```
Comment (s0undt3ch) [inline salt/utils/gitfs.py:2569]:
```suggestion
                change = change or self._remove_cache_dir(
```
So, while iterating the remote hashes, if `change` becomes `True`, then for any of the remaining items in the `remote_set`, `self._remove_cache_dir` will not get called again...

Is this really what you're after?

File: salt/utils/gitfs.py [hunk 3]
```diff
@@ -2604,6 +2630,29 @@ def fetch_remotes(self, remotes=None):
...
+                            repo.get_cache_hash(), branch, ".git", ".salt"
+                        )
+                        fetch_path = salt.utils.path.join(
+                            branch_salt_dir, "fetch_request"
+                        )
+                        if os.path.isdir(branch_salt_dir):
+                            try:
+                                with salt.utils.files.fopen(fetch_path, "w"):
+                                    pass
+                            except Exception as exc:  # pylint: disable=broad-except
```
Comment (s0undt3ch) [inline salt/utils/gitfs.py:2648]:
```suggestion
                            except Exception as exc:  # pylint: disable=broad-except
```
Do we really need this broad exception?

**Review by s0undt3ch — COMMENTED** [review]
You're still planning on adding tests for all of these scenarios that you're trying to cover right?

File: salt/utils/gitfs.py [hunk 4]
```diff
@@ -458,46 +456,29 @@ def __init__(
...
         hash_type = getattr(hashlib, self.opts.get("hash_type", "md5"))
-        # Generate full id.
-        # Full id helps decrease the chances of collections in the gitfs cache.
-        try:
-            target = str(self.get_checkout_target())
-        except AttributeError:
-            target = ""
-        self._full_id = "-".join(
-            [
-                getattr(self, "name", ""),
```
Comment (whytewolf) [inline salt/utils/gitfs.py:469]: I see you are pretty much dropping full_id. however i don't see name being picked up elsewhere. are you still accounting for this functionality?

File: salt/utils/gitfs.py (hunk 2 above)
Comment (cmcmarrow) [reply salt/utils/gitfs.py:2569]: I switch it self._remove ... or change

File: salt/utils/gitfs.py (hunk 4 above)
Comment (cmcmarrow) [reply salt/utils/gitfs.py:469]: yea self._full_id can be completely removed

File: salt/utils/gitfs.py (hunk 2 above)
Comment (cmcmarrow) [reply salt/utils/gitfs.py]: fixed

File: salt/utils/gitfs.py (hunk 4 above)
Comment (cmcmarrow) [reply salt/utils/gitfs.py:469]: yup 100% fine to remove full_id

File: salt/utils/cache.py [hunk 5]
```diff
@@ -345,3 +348,32 @@ def context_cache_wrap(*args, **kwargs):
         return func(*args, **kwargs)
 
     return context_cache_wrap
+
+
+def verify_cache_version(cache_path):
+    """
+    Check that cache version matches salt version.
+    If cache version deos not matches salt version wipe the cache.
```
Comment (twangboy) [inline salt/utils/cache.py:356]: Check that the cached version matches the Salt version.
If the cached version does not match the Salt version, wipe the cache.

File: salt/utils/cache.py [hunk 6]
```diff
@@ -345,3 +348,32 @@ def context_cache_wrap(*args, **kwargs):
...
+
+    :return: True if cache version matched. False if cache version did not match.
```
Comment (twangboy) [inline salt/utils/cache.py:358]:
```
``True`` if cache version matches, otherwise ``False``
```

File: salt/utils/gitfs.py [hunk 7]
```diff
@@ -1274,13 +1289,14 @@ def __init__(
             role,
         )
 
-    def checkout(self):
+    def checkout(self, fetch_on_fail=True):
         """
         Checkout the configured branch/tag. We catch an "Exception" class here
         instead of a specific exception class because the exceptions raised by
         GitPython when running these functions vary in different versions of
         GitPython.
```
Comment (twangboy) [inline salt/utils/gitfs.py:1297]: Can we get some docs on what the `fetch_on_fail` option does?

File: salt/utils/gitfs.py [hunk 8]
```diff
@@ -1602,10 +1627,11 @@ def peel(self, obj):
         except AttributeError:
             return obj.get_object()
 
-    def checkout(self):
+    def checkout(self, fetch_on_fail=True):
         """
         Checkout the configured branch/tag
```
Comment (twangboy) [inline salt/utils/gitfs.py:1632]: Same here... document `fetch_on_fail`

File: salt/utils/gitfs.py [hunk 9]
```diff
@@ -2604,6 +2633,28 @@ def fetch_remotes(self, remotes=None):
...
+                        fetch_path = salt.utils.path.join(
+                            branch_salt_dir, "fetch_request"
+                        )
+                        if os.path.isdir(branch_salt_dir):
+                            try:
+                                with salt.utils.files.fopen(fetch_path, "w"):
+                                    pass
+                            except Exception as exc:  # pylint: disable=broad-except
+                                log.error(
+                                    f"Failed to place fetch request: {fetch_path} {exc}",
```
Comment (twangboy) [inline salt/utils/gitfs.py:2653]: Failed to `make` fetch request...

File: salt/utils/gitfs.py [hunk 10]
```diff
@@ -2604,6 +2633,28 @@ def fetch_remotes(self, remotes=None):
...
+                                    exc_info=True,
+                                )
+                        else:
+                            log.error(f"Failed to place fetch request: {fetch_path}")
```
Comment (twangboy) [inline salt/utils/gitfs.py:2657]: Same here

**Review by twangboy — CHANGES_REQUESTED** [review]
Mostly documentation and customer-facing messages and stuff...

File: tests/pytests/unit/test_minion.py [hunk 11]
```diff
@@ -21,35 +21,33 @@
...
+        assert minion.opts["grains"] == minion_opts["grains"]
         grainsfunc.assert_not_called()
 
 
-def test_minion_load_grains_true():
+def test_minion_load_grains_true(minion_opts):
     """
     Minion generates grains when load_grains is True
     """
-    opts = {"random_startup_delay": 0, "grains": {}}
```
Comment (Ch3LL) [inline tests/pytests/unit/test_minion.py:39]: Can you elaborate on why you are removing these opts from this test and the ones below?

File: salt/utils/gitfs.py [hunk 12]
```diff
@@ -510,11 +494,30 @@ def __init__(
             log.critical(msg, exc_info=True)
             failhard(self.role)
 
-    def full_id(self):
```
Comment (Ch3LL) [inline salt/utils/gitfs.py:513]: We need to properly deprecate a function if its no longer needed.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/test_minion.py (hunk 11 above)
Comment (cmcmarrow) [reply tests/pytests/unit/test_minion.py:39]: minion_opts is a dict with those keys and values init

File: salt/utils/gitfs.py (hunk 12 above)
Comment (cmcmarrow) [reply salt/utils/gitfs.py:513]: It was in the CVE fix

**Comment by cmcmarrow** [comment]
Note: pygit2 wipes locks on fetch

File: tests/pytests/functional/pillar/test_git_pillar.py
Comment (whytewolf) [inline tests/pytests/functional/pillar/test_git_pillar.py:1]: same as below. can we get more it is broken tests. such as bad configs, branches with bad chars in the name such as `/`. branches that are keywords in salt. ect.

File: tests/pytests/functional/utils/test_cache.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        full_path = salt.utils.path.join(tmp_path, full_path)
+        is_file = os.path.isfile(full_path)
+        if ret is None:
+            ret = is_file
+        elif ret is not is_file:
+            return None  # Some files are found and others are missing
+    return ret
+
+
+def test_verify_cache_version(tmp_path):
```
Comment (Ch3LL) [inline tests/pytests/functional/utils/test_cache.py:44]: Can we also ensure we are testing actual Salt versions in the cache file?

File: salt/utils/gitfs.py (hunk 12 above)
Comment (Ch3LL) [reply salt/utils/gitfs.py:513]: If the rest of the team is okay with removing it then I am, but per policy it should properly be deprecated before removal.

File: tests/pytests/unit/test_minion.py (hunk 11 above)
Comment (Ch3LL) [reply tests/pytests/unit/test_minion.py:39]: Yesh, but are these tests relying on non default configs? For example in a test below it had previously set `"return_retry_tries"` to 20 which is not the default.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/utils/gitfs.py (hunk 12 above)
Comment (twangboy) [reply salt/utils/gitfs.py:513]: Remove it... it hasn't had time to have anything built around it...

**Review by twangboy — COMMENTED** [review]
Left my vote on deprecation

File: tests/pytests/unit/test_minion.py (hunk 11 above)
Comment (cmcmarrow) [reply tests/pytests/unit/test_minion.py:39]: yup line 148 changes the default to 20

File: tests/pytests/functional/utils/test_cache.py (hunk 13 above)
Comment (cmcmarrow) [reply tests/pytests/functional/utils/test_cache.py:44]: yes

File: salt/utils/cache.py [hunk 14]
```diff
@@ -345,3 +348,32 @@ def context_cache_wrap(*args, **kwargs):
...
+
+def verify_cache_version(cache_path):
+    """
+    Check that the cached version matches the Salt version.
+    If the cached version does not match the Salt version, wipe the cache.
+
+    :return: ``True`` if cache version matches, otherwise ``False``
+    """
+    if not os.path.isdir(cache_path):
+        os.makedirs(cache_path)
```
Comment (Ch3LL) [inline salt/utils/cache.py:361]: Is this going to create the directory with the correct permissions?

File: tests/pytests/functional/utils/test_cache.py
[the remark this answers is not in the document]
Comment (cmcmarrow) [reply tests/pytests/functional/utils/test_cache.py:1]: only ways cache can break is if the path is invalid, file permissions.
I have a test passing a bad path and another were change the cache to -1

**Review by whytewolf — COMMENTED** [review]
I see no tests for the name functionality. I know i had asked for that before. since the name functionality had changed when full_id was introduced and doesn't look to be readded but the functionality is still documented it needs to be checked and made sure to exist and function. there needs to be tests that test the full range of options not just straight branch repo lines. 

also I am seeing tests that assert `_` is a valid cache basename?

**Review by Ch3LL — COMMENTED** [review]
Just one more question

**Comment by cmcmarrow** [comment]
@whytewolf yes repos that are not `__env__` will default to `_`.

**Comment by whytewolf** [comment]
> 
wait that makes no sense. the REPO will be named `_` if the BRANCH is not `__env__`? that will break everyone not using `__env__` that has multiple repos setup. please tell me you meant the branch will default to `_`. which is still not ideal. I would much prefer base. or the actual branch name. or what ever saltenv is for the branch.

**Comment by cmcmarrow** [comment]
@whytewolf here is an example:

- main https://github.com/saltstack/salt -> CJSDKJAHEBWS/_
- https://github.com/saltstack/salt -> JSKAJFUWISKWS/_
- https://github.com/saltstack/other_repo -> AHDJWHSHW/_
- __env__ https://github.com/saltstack/salt -> ALKDFHJSJHSAB/master

pillarenv=main

- main https://github.com/saltstack/salt -> CJSDKJAHEBWS/_
- https://github.com/saltstack/salt -> JSKAJFUWISKWS/_
- https://github.com/saltstack/other_repo -> AHDJWHSHW/_
- __env__ https://github.com/saltstack/salt -> ALKDFHJSJHSAB/main

All four get different hashes and __env__ switches dir
We only nane the __env__ because that's should be the only dynamically changing repo
GitFS does not provide self.branch which makes self.get_checkout_target() fail for GitFS. Ontop GitFS does not run checkout. So it does not need a branch on init because you can give your branch in file_list({"saltenv": "main"}) which does not populate the cache just lists the files.

Before my changes repos only got a hash but that was a problem for pillar which uses `__env__`

**Comment by whytewolf** [comment]
> @whytewolf here is an example:
> 
>     * main https://github.com/saltstack/salt -> CJSDKJAHEBWS/_
> 
>     * https://github.com/saltstack/salt -> JSKAJFUWISKWS/_
> 
>     * https://github.com/saltstack/other_repo -> AHDJWHSHW/_
> 
>     * **env** https://github.com/saltstack/salt -> ALKDFHJSJHSAB/master
> 
> 
> pillarenv=main
> 
>     * main https://github.com/saltstack/salt -> CJSDKJAHEBWS/_
> 
>     * https://github.com/saltstack/salt -> JSKAJFUWISKWS/_
> 
>     * https://github.com/saltstack/other_repo -> AHDJWHSHW/_
> 
>     * **env** https://github.com/saltstack/salt -> ALKDFHJSJHSAB/main
> 
> 
> All four get different hashes and **env** switches dir We only nane the **env** because that's should be the only dynamically changing repo GitFS does not provide self.branch which makes self.get_checkout_target() fail for GitFS. Ontop GitFS does not run checkout. So it does not need a branch on init because you can give your branch in file_list({"saltenv": "main"}) which does not populate the cache just lists the files.
> 
> Before my changes repos only got a hash but that was a problem for pillar which uses `__env__`

ok, that makes sense, and is safer than it originally sounded, so the repo is still a hash and the ref is `_`. we do still need name. which would replace the repo hash. so instead of `hash/_` would be `name/_` this allows users to have multiple of the same repo that might say mount in different places. 

i was concerned because the labeling on the function makes it sound like repo is `_` not the env.

**Comment by cmcmarrow** [comment]
@whytewolf can you please give an example where two non `__env__` repos can collided

Here you can see they get different hashes. The hash gets made from the id.
 * `main https://github.com/saltstack/salt` -> `CJSDKJAHEBWS/_`
 * `https://github.com/saltstack/salt` -> `JSKAJFUWISKWS/_`

non `__env__` repos do not change branches. Even if you change `pillarenv` or `saltenv`. There branch is in `self.branch` which comes from `get_url()`

`GitFS` cant even get a branch for checkout
`Winrepo` and `pillar` can get a branch from two places
* From id `main https://github.com/saltstack/salt` which is static
* From opts `{self._role}_branch` which is static


`GitProvider.get_url()` # makes `self.branch` when `winrepo` or `pillar`
`GitProvider.get_checkout_target()` # uses `self.branch` unless `__env__` then we look at `pillarenv` or `saltenv`
`GitProvider.checkout()` <- # populates cache dir with files from the checkout target

I would keep the name `_` as for `GitFS` does not checkout a branch. It will fetch repo data and look at multiple branches with out writing to the cachedir. Changing the `_` to current `saltenv` would make things confusing because that repo can be used for multiple branches.

**Comment by whytewolf** [comment]
> @whytewolf can you please give an example where two non `__env__` repos can collided
> 
> Here you can see they get different hashes. The hash gets made from the id.
> 
>     * `main https://github.com/saltstack/salt` -> `CJSDKJAHEBWS/_`
> 
>     * `https://github.com/saltstack/salt` -> `JSKAJFUWISKWS/_`
> 
> 
> non `__env__` repos do not change branches. Even if you change `pillarenv` or `saltenv`. There branch is in `self.branch` which comes from `get_url()`
> 
> `GitFS` cant even get a branch for checkout `Winrepo` and `pillar` can get a branch from two places
> 
>     * From id `main https://github.com/saltstack/salt` which is static
> 
>     * From opts `{self._role}_branch` which is static
> 
> 
> `GitProvider.get_url()` # makes `self.branch` when `winrepo` or `pillar` `GitProvider.get_checkout_target()` # uses `self.branch` unless `__env__` then we look at `pillarenv` or `saltenv` `GitProvider.checkout()` <- # populates cache dir with files from the checkout target
> 
> I would keep the name `_` as for `GitFS` does not checkout a branch. It will fetch repo data and look at multiple branches with out writing to the cachedir. Changing the `_` to current `saltenv` would make things confusing because that repo can be used for multiple branches at once.


what is the result of this? this does break even though they are the same repo because of the locking. and is the reason that name was introduced.

```
main https://github.com/saltstack/salt:
  root: salt/modules
  mount: _modules
main https://github.com/saltstack/salt:
  root: salt/states
  mount: _states
```

File: salt/utils/cache.py (hunk 14 above)
Comment (cmcmarrow) [reply salt/utils/cache.py:361]: `os.mkdirs` is used all over gitfs so I believe this is the right way for gitfs.

File: tests/pytests/functional/pillar/test_git_pillar.py
Comment (dmurphy18) [reply tests/pytests/functional/pillar/test_git_pillar.py:1]: @whytewolf Currently trying to get enough tests to ensure it is working and allow for the 3006.3 release.
We can do additional tests, including bad configs, etc. for a later release, e.g. 3006.4, especially as part of increasing code coverage on the gitfs handling code.

But for now we need to get the release out.

File: tests/pytests/functional/utils/test_cache.py
[the remark this answers is not in the document]
Comment (dmurphy18) [reply tests/pytests/functional/utils/test_cache.py:1]: see comment above about sufficient testing to get a release out and add additional testing during increasing code coverage.

**Review by dmurphy18 — COMMENTED** [review]
Left comments for @whytewolf

**Review by twangboy — DISMISSED** [review]

**Comment by whytewolf** [comment]
@cmcmarrow ok. there is a name issue. 

```
2023-09-05 17:35:44,308 [salt.utils.gitfs :792 ][DEBUG   ][3907641] Current fetch URL for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git': None (desired: git@github.com:whytewolf/salt-phase0-pillar.git)
2023-09-05 17:35:44,308 [salt.utils.gitfs :801 ][DEBUG   ][3907641] Fetch URL for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git' set to git@github.com:whytewolf/salt-phase0-pillar.git
2023-09-05 17:35:44,308 [salt.utils.gitfs :818 ][DEBUG   ][3907641] Current refspecs for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git': [] (desired: ['+refs/heads/*:refs/remotes/origin/*', '+refs/tags/*:refs/tags/*'])
2023-09-05 17:35:44,308 [salt.utils.gitfs :827 ][DEBUG   ][3907641] Refspecs for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git' set to ['+refs/heads/*:refs/remotes/origin/*', '+refs/tags/*:refs/tags/*']
2023-09-05 17:35:44,308 [salt.utils.gitfs :844 ][DEBUG   ][3907641] Current http.sslVerify for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git': None (desired: true)
2023-09-05 17:35:44,309 [salt.utils.gitfs :853 ][DEBUG   ][3907641] http.sslVerify for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git' set to true
2023-09-05 17:35:44,309 [salt.utils.gitfs :866 ][DEBUG   ][3907641] Config updates for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git' written to /var/cache/salt/master/git_pillar/vYsBEzVHxGMN4R9eR4zrIKhGEJyLt8SdJ4E7xhrGoWM=/_/.git/config
2023-09-05 17:35:44,311 [salt.utils.gitfs :792 ][DEBUG   ][3907641] Current fetch URL for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git': git@github.com:whytewolf/salt-phase0-pillar.git (desired: git@github.com:whytewolf/salt-phase0-pillar.git)
2023-09-05 17:35:44,311 [salt.utils.gitfs :818 ][DEBUG   ][3907641] Current refspecs for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git': ['+refs/heads/*:refs/remotes/origin/*', '+refs/tags/*:refs/tags/*'] (desired: ['+refs/heads/*:refs/remotes/origin/*', '+refs/tags/*:refs/tags/*'])
2023-09-05 17:35:44,311 [salt.utils.gitfs :844 ][DEBUG   ][3907641] Current http.sslVerify for git_pillar remote 'master git@github.com:whytewolf/salt-phase0-pillar.git': true (desired: true)
2023-09-05 17:35:44,311 [salt.utils.gitfs :2544][CRITICAL][3907641] The following git_pillar remotes have conflicting cachedirs: master git@github.com:whytewolf/salt-phase0-pillar.git, master git@github.com:whytewolf/salt-phase0-pillar.git. Resolve this using a per-remote parameter called 'name'.
```

```
  - git:
    - master git@github.com:whytewolf/salt-phase0-pillar.git:
      - name: testing
    - master git@github.com:whytewolf/salt-phase0-pillar.git:
      - name: set2
      - mountpoint: test/
```

I used name in both places which is meant to avoid the above error. this also happens with `__env__`

**Review by whytewolf — APPROVED** [review]

**Review by dmurphy18 — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Comment by tysonnorris** [comment]
A question on this - is it safe to use multiple gitfs branches/envs on same salt-master BEFORE this change if we set `gitfs_global_lock: False`?  

I assume it is safe AFTER this change, especially if the locks move to salt working dir.

**Pull request merged**